### PR TITLE
Sina/add joystick2

### PR DESCRIPTION
**Pull request by sinaaghli**
Joystick control carrot  pose

### CODE REVIEW

File: ocs2_ros_interfaces/src/command/TargetTrajectoriesKeyboardPublisher.cpp [hunk 1]
```diff
@@ -84,6 +93,48 @@ void TargetTrajectoriesKeyboardPublisher::publishKeyboardCommand(const std::stri
...
 
+/******************************************************************************************************/
+/******************************************************************************************************/
+/******************************************************************************************************/
+void TargetTrajectoriesKeyboardPublisher::publishJoystickCommand(const std::string& commadMsg) {
+  while (rclcpp::ok()) {
+    std::cout << commadMsg << ": ";
+    std::this_thread::sleep_for(5s);
+    // getCommandLine();
+    // vector_t commandLineInput = getCommandLine().cwiseMin(targetCommandLimits_).cwiseMax(-targetCommandLimits_);
```
Comment (sinaaghli) [inline ocs2_ros_interfaces/src/command/TargetTrajectoriesKeyboardPublisher.cpp]: @kylc @senceryazici if I don't call this function getCommandLine() 
then the process dies for some reason!

**Comment by mamoll** [comment]
There seem to be quite a few arbitrary unrelated whitespace changes. Merging them might complicate rebasing later on. 

Also, there are still a bunch of print statements you'll want to remove (or replace ROS logging if they need to stay).

File: ocs2_robotic_examples/ocs2_legged_robot_ros/CMakeLists.txt [hunk 2]
```diff
@@ -3,9 +3,11 @@ project(ocs2_legged_robot_ros)
```
Comment (kylc) [inline ocs2_robotic_examples/ocs2_legged_robot_ros/CMakeLists.txt]: Lots of unrelated changes in this file, I'm not sure if anything actually changed? Can you revert the formatting noise?

File: ocs2_robotic_examples/ocs2_legged_robot_ros/launch/legged_robot_ddp.launch.py [hunk 3]
```diff
@@ -76,6 +76,12 @@ def generate_launch_description():
             output='screen',
             prefix=["gnome-terminal --"],
         ),
+        Node(
+            name='joy_node',
+            package='joy',
+            executable='joy_node',
+            output='screen',
+        ),
```
Comment (kylc) [inline ocs2_robotic_examples/ocs2_legged_robot_ros/launch/legged_robot_ddp.launch.py:84]: Just checking, does this behave properly if I don't have a joystick plugged in? Or should it be gated behind a `joystick:=true` launch parameter?

**Comment by sinaaghli** [comment]
> There seem to be quite a few arbitrary unrelated whitespace changes. Merging them might complicate rebasing later on.
> 
> Also, there are still a bunch of print statements you'll want to remove (or replace ROS logging if they need to stay).

Thanks @mamoll , I forgot about the couts! and I didn't realize my vscode auto-formatted the cmake file! they should be reverted now!

**Review by kylc — CHANGES_REQUESTED** [review]
It seems like GitHub isn't showing the true diff. It's only showing a couple of lines in the .h and .cpp files changed in total. Can you see what's going on?

EDIT: Seems the diff is showing correctly now, I will take a closer look.

File: ocs2_ros_interfaces/src/command/TargetTrajectoriesJoystickPublisher.cpp [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+  }  // end of while loop
+}
+
+void TargetTrajectoriesJoystickPublisher::joyCallback(const sensor_msgs::msg::Joy::SharedPtr msg) 
+{
+  const std::lock_guard<std::mutex> lock(joymutex_);
+  command_(0) = msg->axes[1];
+  command_(1) = msg->axes[0];
+  command_(2) = 0.0;
+  command_(3) = msg->axes[3]*90;
```
Comment (kylc) [inline ocs2_ros_interfaces/src/command/TargetTrajectoriesJoystickPublisher.cpp]: `90` magic number?

File: ocs2_ros_interfaces/src/command/TargetTrajectoriesJoystickPublisher.cpp [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        std::lock_guard<std::mutex> lock(latestObservationMutex_);
+        observation = latestObservation_;
+      }
+
+      // Rotate joystick command to world frame since that's what ocs2 expects
+      Eigen::Quaterniond curr_rot = Eigen::AngleAxisd(observation.state[9],Eigen::Vector3d::UnitZ())
+                                * Eigen::AngleAxisd(observation.state[10],Eigen::Vector3d::UnitY())
+                                * Eigen::AngleAxisd(observation.state[11],Eigen::Vector3d::UnitX());
+      joy_command.segment<3>(0) = curr_rot.matrix()*joy_command.segment<3>(0);
+      const vector_t commandLineInput(joy_command);
```
Comment (kylc) [inline ocs2_ros_interfaces/src/command/TargetTrajectoriesJoystickPublisher.cpp]: Maybe consider fixing some of these names, because it no longer comes from the command line.

File: ocs2_ros_interfaces/src/command/TargetTrajectoriesJoystickPublisher.cpp [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+      {
+        std::lock_guard<std::mutex> lock(latestObservationMutex_);
+        observation = latestObservation_;
+      }
+
+      // Rotate joystick command to world frame since that's what ocs2 expects
+      Eigen::Quaterniond curr_rot = Eigen::AngleAxisd(observation.state[9],Eigen::Vector3d::UnitZ())
+                                * Eigen::AngleAxisd(observation.state[10],Eigen::Vector3d::UnitY())
+                                * Eigen::AngleAxisd(observation.state[11],Eigen::Vector3d::UnitX());
+      joy_command.segment<3>(0) = curr_rot.matrix()*joy_command.segment<3>(0);
```
Comment (kylc) [inline ocs2_ros_interfaces/src/command/TargetTrajectoriesJoystickPublisher.cpp:92]: Do we necessarily want to apply the full rotation? I think only rotating about the yaw axis would be sufficient.

Otherwise, if your body is pitched up/down for example (tracking error, holding a pose, or something else) then the desired walking path would ascend/descend off the ground.

Or does OCS2 handle all that for us internally?

File: ocs2_robotic_examples/ocs2_legged_robot_ros/src/LeggedRobotPoseCommandNode.cpp [hunk 7]
```diff
@@ -123,10 +124,10 @@ int main(int argc, char* argv[]) {
 
   // goalPose: [deltaX, deltaY, deltaZ, deltaYaw]
   const scalar_array_t relativeBaseLimit{10.0, 10.0, 0.2, 360.0};
-  TargetTrajectoriesKeyboardPublisher targetPoseCommand(nodeHandle, robotName, relativeBaseLimit, &commandLineToTargetTrajectories);
+  TargetTrajectoriesJoystickPublisher targetPoseCommand(nodeHandle, robotName, relativeBaseLimit, &commandLineToTargetTrajectories);
 
   const std::string commandMsg = "Enter XYZ and Yaw (deg) displacements for the TORSO, separated by spaces";
-  targetPoseCommand.publishKeyboardCommand(commandMsg);
+  targetPoseCommand.publishJoystickCommand(commandMsg);
```
Comment (kylc) [inline ocs2_robotic_examples/ocs2_legged_robot_ros/src/LeggedRobotPoseCommandNode.cpp]: Can we switch between these on a launch option, so we can still use the keyboard if needed? I don't always have my joystick with me :P

**Review by kylc — CHANGES_REQUESTED** [review]

File: ocs2_robotic_examples/ocs2_legged_robot_ros/launch/legged_robot_ddp.launch.py (hunk 3 above)
Comment (sinaaghli) [reply ocs2_robotic_examples/ocs2_legged_robot_ros/launch/legged_robot_ddp.launch.py:84]: if a joystick is not attached then nothing happens, as soon as it gets plugged in then it publishes the appropriate commands

File: ocs2_ros_interfaces/src/command/TargetTrajectoriesJoystickPublisher.cpp (hunk 4 above)
Comment (sinaaghli) [reply ocs2_ros_interfaces/src/command/TargetTrajectoriesJoystickPublisher.cpp]: yes, just put it so that it turns fast enough when joystick is pushed all the way!

File: ocs2_ros_interfaces/src/command/TargetTrajectoriesJoystickPublisher.cpp (hunk 6 above)
Comment (sinaaghli) [reply ocs2_ros_interfaces/src/command/TargetTrajectoriesJoystickPublisher.cpp:92]: that's fair, the command is relative to current pose so I'll just drop the Z component!
I'm not sure we want to add new buttons to also modify the z-axis!

File: ocs2_robotic_examples/ocs2_legged_robot_ros/src/LeggedRobotPoseCommandNode.cpp (hunk 7 above)
Comment (sinaaghli) [reply ocs2_robotic_examples/ocs2_legged_robot_ros/src/LeggedRobotPoseCommandNode.cpp]: sure!

File: ocs2_robotic_examples/ocs2_legged_robot_ros/src/LeggedRobotPoseCommandNode.cpp [hunk 8]
```diff
@@ -123,10 +124,21 @@ int main(int argc, char* argv[]) {
 
   // goalPose: [deltaX, deltaY, deltaZ, deltaYaw]
   const scalar_array_t relativeBaseLimit{10.0, 10.0, 0.2, 360.0};
-  TargetTrajectoriesKeyboardPublisher targetPoseCommand(nodeHandle, robotName, relativeBaseLimit, &commandLineToTargetTrajectories);
 
   const std::string commandMsg = "Enter XYZ and Yaw (deg) displacements for the TORSO, separated by spaces";
-  targetPoseCommand.publishKeyboardCommand(commandMsg);
+  
+  bool use_joystick;
+  nodeHandle->get_parameter("use_joystick", use_joystick);
```
Comment (sinaaghli) [inline ocs2_robotic_examples/ocs2_legged_robot_ros/src/LeggedRobotPoseCommandNode.cpp]: @kylc this was my attempt at reading a parameter to either use the keyboard or joystick, but its not working unfortunetaly, I think since the node is not spinning all the time and when its blocked to keyboard command then its not able to listen to parameter change?

File: ocs2_robotic_examples/ocs2_legged_robot_ros/CMakeLists.txt (hunk 2 above)
Comment (sinaaghli) [reply ocs2_robotic_examples/ocs2_legged_robot_ros/CMakeLists.txt]: I think you were looking at outdated commits!

File: ocs2_ros_interfaces/src/command/TargetTrajectoriesJoystickPublisher.cpp (hunk 4 above)
Comment (kylc) [reply ocs2_ros_interfaces/src/command/TargetTrajectoriesJoystickPublisher.cpp]: If the joystick axis is on the range [-1, 1] can we extract this as a constant called "max yaw rate" or something?

File: ocs2_robotic_examples/ocs2_legged_robot_ros/src/LeggedRobotPoseCommandNode.cpp (hunk 8 above)
Comment (kylc) [reply ocs2_robotic_examples/ocs2_legged_robot_ros/src/LeggedRobotPoseCommandNode.cpp]: It doesn't have to be a ROS parameter. You could load a parameter from the OCS2 config file, like they do just a couple of lines above this.

https://github.com/PickNikRobotics/ocs2/blob/e4aaaa696a0b851094e2bb461a769b661d2cbf1e/ocs2_robotic_examples/ocs2_legged_robot_ros/src/LeggedRobotPoseCommandNode.cpp#L123

**Comment by sinaaghli** [comment]
@kylc  I think I covered your fb in the [last commit](https://github.com/PickNikRobotics/ocs2/pull/3/commits/cf8a831af416fdf29662e0d0ca01b755b2cb6745)

let me know if I can make more updates!

**Review by kylc — APPROVED** [review]
Added height control, look good to me.
